Per detector placzek and vanadium

DanielMurphy22 opened the pull request:
**Description of work.**

Implement a per-detector Vanadium correction in the ISIS Powder Diffraction scripts. This is motivated primarily by the POLARIS total scattering data reduction where it's desirable to have the option to apply the Placzek correction on a per detector basis.

I've added a couple of reference .nxs files for system tests. These are quite large (~200MB). I looked into why they're larger than the .nxs files on the archive that are only ~20MB and there are a few differences:
- the archive files use an ISIS raw nexus format
- the archive files store integer counts as opposed to the floating point values in the reduced files (factor of 2) and the integer data seems to achieve a higher compression ratio (factor of 3-4)
- I don't think the archive files have any errors in them so this halves the size (the errors and the  y values seem to take up most space)

Possible further work:
- the function `divide_by_number_of_detectors_in_bank` is currently duplicated in focus.py and the workflow algorithm `TotScatCalculateSelfScattering`. Hard to centralise this with one of the calls sitting inside a workflow algorithm
- add a flag to control whether Placzek correction is applied during a per detector focus - or could just add filter to apply it when mode='PDF'. For POLARIS it always gets applied right now which might not be what's needed if they just wanted to do a per detector normalisation on a powder sample

**To test:**

Run the following per detector script in MantidWorkbench:
``` python
from mantid.simpleapi import *
import numpy as np
from isis_powder import polaris, SampleDetails
import time

start_time = time.time()
config_file_path = r".../polaris-calculate-pdf/polaris_config_example.yaml"
polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF", per_detector_vanadium=True)

sample_details = SampleDetails(height=4.0, radius=0.2985, center=[0, 0, 0], shape='cylinder')
sample_details.set_material(chemical_formula='Si', packing_fraction=0.6)
sample_details.set_container(radius=0.3175, chemical_formula='V')
polaris.set_sample_details(sample=sample_details)

polaris.create_vanadium(first_cycle_run_no="98532", multiple_scattering=False)
polaris.focus(run_number="98533", input_mode='Summed', sample_empty="98531")
polaris.create_total_scattering_pdf(run_number="98533", merge_banks=False)
  
print('\n\n')
print("Script run time", time.time() - start_time)
```

Run the following per bank (main workflow) script in MantidWorkbench:
``` python
from mantid.simpleapi import *
import numpy as np
from isis_powder import polaris, SampleDetails
import time

start_time = time.time()
config_file_path = r".../polaris-calculate-pdf/polaris_config_example.yaml"
polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF"))

sample_details = SampleDetails(height=4.0, radius=0.2985, center=[0, 0, 0], shape='cylinder')
sample_details.set_material(chemical_formula='Si', packing_fraction=0.6)
sample_details.set_container(radius=0.3175, chemical_formula='V')
polaris.set_sample_details(sample=sample_details)

polaris.create_vanadium(first_cycle_run_no="98532", multiple_scattering=False)
polaris.focus(run_number="98533", input_mode='Summed', sample_empty="98531")
polaris.create_total_scattering_pdf(run_number="98533", merge_banks=False)
  
print('\n\n')
print("Script run time", time.time() - start_time)
```
Both should give similar output PDF_r plots

Fixes #32919
Supercedes old PR: #33008

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DanielMurphy22 committed:
Allow ranges in SplineBackground

Calculated Y values are the same as before
Note the outputWS SharedX values are from the currentWSIndex rather 0

DanielMurphy22 committed:
Add calibrate per detector methods

DanielMurphy22 committed:
Add create_van_per_detector method

DanielMurphy22 committed:
Support saving unsplined MatrixWorkspaces aswell as groups

DanielMurphy22 committed:
Add tests for current focus

DanielMurphy22 committed:
restructure half focus

DanielMurphy22 committed:
Fix Focus tests to genuinely use per_detector

ConvertFromDistribution for per detector before focussing

DanielMurphy22 committed:
set per_detector consistently

DanielMurphy22 committed:
Rewrite tests to not workaround bugfix

DanielMurphy22 committed:
Make per detector routine output as Distribution

Also enforce placzek_run_number input when using per_detector routine

DanielMurphy22 committed:
Pass PaalmanPingsFocus Test

DanielMurphy22 committed:
Edit TotScattCalculateSelfScattering and polaris_algs for per_detector

DanielMurphy22 committed:
Rework TotScattCalculateSelfScattering+ polaris_algs for per_detector

DanielMurphy22 committed:
Remove monitor and dummy workspaces

Allows correct RebinForWorkspace before applying vanadium normalization

DanielMurphy22 committed:
Remove unneccessary file

DanielMurphy22 committed:
Fix create pdf for per bank

DanielMurphy22 committed:
Rename function

DanielMurphy22 committed:
remove tests and dummy function

DanielMurphy22 committed:
Put spline_coefficient_per_detector in mapping file

DanielMurphy22 committed:
Convert to TOF after applying corrections

DanielMurphy22 committed:
Run convertToDistribution and FromDistribution both in dSpacing

DanielMurphy22 committed:
Wrap ConvertToFromDistribution in function to ensure in dSpacing

DannyHindson committed:
Apply black formatting

DannyHindson committed:
Fix up following rebase - mostly black formatting

DannyHindson committed:
Fix doxygen problems

DannyHindson committed:
Delete grouping workspace after it's used in Divide

DannyHindson committed:
Add release notes

DannyHindson committed:
Various changes

Move the Placzek correction code into the polaris.py and polaris_algs.py
units
Pull out absolute normalisation code into function to remove duplication
Stop TotScatCalculateSelfScattering deleting the input workspace
Remove the Placzek_run_number parameter. Seems like this will always be
the same as the run number
If not doing any V normalisation follow the same code path after focussing
as the "per detector" V normalisation workflow ie divide by number of detectors
in each bank
Keep the ADS name the same during some algorithm calls to avoid unintentional
intermediate workspaces being left behind

DannyHindson committed:
Add default value to new parameter so other instruments unaffected

DannyHindson committed:
Add new per detector parameters to all instruments

Also rename new per detector parameter

DannyHindson committed:
Tidy up some more intermediate workspaces

DannyHindson committed:
If vanadium_normalisation=false leave result with distribution=false

Fixes PEARL system tests running with vanadium_normalisation=false
Since there's now a difference between the code path for
vanadium_normalisation=false and the per detector version of
vanadium_normalisation=true, only divide by number of detectors in the
bank in the per detector case.
Doesn't seem much benefit in altering the vanadium_normalisation=false
behaviour for the per bank process any more

DannyHindson committed:
Improve convert_between_distribution function

Use UnitID instead of caption to remove mapping logic

DannyHindson committed:
Rework the per detector V Bragg peak masking logic

The previous logic relied on the bank names being bank1, bank2 etc
Not the case for HRPD so new logic more generic
Move code into abstract_inst so it can be overridden per instrument
Also remove monitor spectra from POLARIS unsplined per detector
V output so number of spectra same as the splined version

DannyHindson committed:
Add per detector system tests

Also clean up all files properly - especially because the per detector
files are quite large

DannyHindson committed:
Add documentation of per detector V option

RichardWaiteSTFC commented:
I'm getting this error from the `init` of `Polaris`
```
ValueError: Unknown configuration key: chopper_on
  File "C:/Users/xhg73778/OneDrive - Science and Technology Facilities Council/test_polaris.py", line 8, in <module>
    polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF", per_detector=True)
  File "C:\mantid-conda\mantid\scripts\Diffraction\isis_powder\polaris.py", line 18, in __init__
    self._inst_settings = instrument_settings.InstrumentSettings(
  File "C:\mantid-conda\mantid\scripts\Diffraction\isis_powder\routines\instrument_settings.py", line 44, in __init__
    self._parse_attributes(dict_to_parse=self._basic_conf_dict)
  File "C:\mantid-conda\mantid\scripts\Diffraction\isis_powder\routines\instrument_settings.py", line 111, in _parse_attributes
    raise ValueError("Unknown configuration key: " + str(config_key))
```
I'm using the config file here
`scripts/Diffraction/isis_powder/polaris_routines/Examples/polaris_config_example.yaml`
Which does indeed have the line
`chopper_on : False`

Is there a different file I should be using?

DannyHindson commented:
The `chopper_on` setting was renamed to `mode` a while ago (in https://github.com/mantidproject/mantid/pull/19761) but it looks like this example file was missed. So if you change this line in the .yaml file to `mode="PDF"` it should work. On recent POLARIS PRs we've been using a set of files that's been passed around me\Harriet\Dan (see https://github.com/mantidproject/mantid/pull/34144) that includes the latest settings but I see we forgot to put that link on this PR.

I can update this example file as part of the review changes for this PR.

RichardWaiteSTFC commented:
> I can update this example file as part of the review changes for this PR.

Thanks, I think the only reference in the code is here
https://github.com/mantidproject/mantid/blob/1a196143928bdc23c7685b12db75a3c51bd7fd11/docs/source/algorithms/MatchAndMergeWorkspaces-v1.rst
but it would probably be good to update it anyway

RichardWaiteSTFC commented:
FYI from the config keys in `polaris_param_mapping.py` I think the kwarg should be `per_detector_vanadium=True)` (`per_detector` gives a `ValueError`). I will update the testing instructiuons.

RichardWaiteSTFC commented on scripts/Diffraction/isis_powder/routines/common.py:
```diff
@@ -707,3 +712,44 @@ def workspace_has_current(ws):
+            detectors_to_mask.append(cal_workspace.getDetectorIDs(wsIndex)[0])
+
+    # Remove Masked and Monitor spectra
+    correction_workspace = mantid.ExtractMonitors(
```
Could some of this be done in a loop e.g.
```
for ws in [correction_workspace, data_workspace]:
    # remove spectra
```

DannyHindson replied:
The same .cal file is used to mask both workspaces so I think the spectra that get masked should be the same?

RichardWaiteSTFC replied:
Yep that's good, as long as we don't go through the work of getting the masked detectors twice form the same file - which i don't think it does, looks like it's done before the new loop!

RichardWaiteSTFC commented on Framework/CurveFitting/src/Algorithms/SplineBackground.cpp:
```diff
@@ -34,6 +34,7 @@ void SplineBackground::init() {
   mustBePositive->setLower(0);
   declareProperty("WorkspaceIndex", 0, mustBePositive, "The index of the spectrum for fitting.");
   declareProperty("NCoeff", 10, "The number of b-spline coefficients.");
+  declareProperty("EndWorkspaceIndex", 0, mustBePositive, "The end index of the spectrum range for fitting.");
```
It doesn't look like this new parameter has a unit test associated with it?

RichardWaiteSTFC commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/TotScatCalculateSelfScattering.py:
```diff
@@ -56,12 +56,15 @@ def PyInit(self):
             defaultValue="",
             doc="Sample Temperature in Kelvin. Required for 2nd order Placzek correction if not " "using Sample Logs.",
         )
+        self.declareProperty(name="ApplyPerDetector", defaultValue=False, doc="Apply the correction to unfocussed data.")
```
It seems there is a unit test for this file, but it's a bit light. I think given the nature of this algorithm a systems test covers this.

RichardWaiteSTFC commented on scripts/Diffraction/isis_powder/routines/focus.py:
```diff
@@ -151,18 +156,106 @@ def _focus_one_ws(
+    return processed_workspace
+
+
+def divide_by_number_of_detectors_in_bank(focussed_data, cal_filepath):
```
Is this a (near?) duplicate of what is defined in the algorithm `TotScatCalculateSelfScattering`? could be defined there and imported here?

RichardWaiteSTFC reviewed (changes requested):
Nice feature thanks for finishing this off. A couple of thoughts for future (probably to me!):

1. It would have been nice to generalise the `per_bank` and `per_detectors` workflows to some extent - for example perhaps there could have been more common code if in the bank case had a workspace with a spectrum from each bank as opposed to a group workspace of individual spectra.
2. I think there should be a common instrument `param_mapping.py` file for common config flags. Maybe a nice new starter issue?

But I think the above is definitely for another day/low priority!

Other than that, just some minor suggestions, but only if you have time! If not I'm happy to approve!

DannyHindson commented on scripts/Diffraction/isis_powder/polaris_routines/polaris_algs.py:
```diff
@@ -202,22 +193,91 @@ def _obtain_focused_run(run_number, focus_file_path):
+        CalFileName=run_details.grouping_file_path,
+        SampleGeometry=sample_geometry,
+        SampleMaterial=sample_material,
+        CrystalDensity=sample_details.material_object.number_density_effective,
```
The `CrystalDensity` parameter was removed from `TotScatCalculateSelfScattering` in ~July last year so it shouldn't be used on either call now. Perhaps due to the fact this PR was started a long time ago and then paused. I'll remove

DannyHindson commented on Framework/CurveFitting/src/Algorithms/SplineBackground.cpp:
```diff
@@ -182,15 +196,16 @@ double SplineBackground::calculateBinWeight(double errValue) {
-                    " unmasked bin. This bin will have no weight during the fitting"
-                    " process");
-    }
+    // Temporarily removed this as the warnings slow the algorithm immensely
```
Hadn't spotted this. Must have been something Dan did. In other algorithms I've seen a check like this. Perhaps this might help here:
```
    if (g_log.is(Kernel::Logger::Priority::PRIO_DEBUG)) {
      g_log.debug("message");
    }
```

DannyHindson replied:
I noticed that the call to SplineBackground from the function \_spline_vanadium_ws_per_detector in abstract_inst.py uses EnableLogging=False so maybe the commenting out isn't needed any more at all. I'll add my suggested if statement in anyway to protect any users who are running SplineBackground across multiple spectra from places other than these POLARIS scripts

DannyHindson committed:
Apply review comments

Support 2nd order placzek correction if specified
Only apply placzek per detector correction if mode=PDF
Add unit test for new SplineBackground parameter
Only override the pdf type and placzek order with defaults if they're
not in POLARIS object attributes (as well as not in the kwargs of the calls
to focus or create_total_scattering_pdf)

DannyHindson commented on scripts/Diffraction/isis_powder/routines/common.py:
```diff
@@ -707,3 +712,44 @@ def workspace_has_current(ws):
+    )
+
+    detectors_to_mask = []
+    for wsIndex in range(0, cal_workspace.getNumberHistograms()):
```
So would I need to do other changes in addition to changing `dataY` to `extractY` to get this vectorized?

RichardWaiteSTFC replied:
So this looks like it contains a single bin/value per spectrum - so in this case `extractY()` will return a vector with these values so you can do something like `np.where(cal_workspace.extractY())` - i think!

DannyHindson committed:
Update incorrect setting in example POLARIS config

RichardWaiteSTFC commented on scripts/Diffraction/isis_powder/polaris.py:
```diff
@@ -75,13 +75,13 @@ def ensure_per_detector_and_vanadium_output_are_in_sync(self, vanadium_d, per_de
-        if "pdf_type" not in kwargs or kwargs["pdf_type"] not in ["G(r)", "g(r)", "RDF(r)", "G_k(r)"]:
-            kwargs["pdf_type"] = "G(r)"
+        self._inst_settings.update_attributes(kwargs=kwargs)
+        if self._inst_settings.pdf_type is None or self._inst_settings.pdf_type not in ["G(r)", "g(r)", "RDF(r)", "G_k(r)"]:
```
I think this is working properly now. Before when I ran the test script I didn't need to specify the `pdf_type` in the function `create_total_scattering_pdf` even though it wasn't in the config file (it defaulted to `G(r)`). Now it throws an error.

DannyHindson replied:
Do you get an error or a warning? If no pdf type is specified it should default it to G(r) and provide a warning

RichardWaiteSTFC replied:
Ah sorry I got an error:
```
AttributeError: The parameter with name: 'pdf_type' is required but was not set or passed.

 File "C:/Users/xhg73778/OneDrive - Science and Technology Facilities Council/test_polaris.py", line 17, in <module>
 pol.create_total_scattering_pdf(run_number="98533", merge_banks=False)
 File "C:\mantid-conda\mantid\scripts\Diffraction\isis_powder\polaris.py", line 79, in create_total_scattering_pdf
 if self._inst_settings.pdf_type is None or self._inst_settings.pdf_type not in ["G(r)", "g(r)", "RDF(r)", "G_k(r)"]:
 File "C:\mantid-conda\mantid\scripts\Diffraction\isis_powder\routines\instrument_settings.py", line 62, in __getattr__
 self._raise_user_param_missing_error(known_param)
 File "C:\mantid-conda\mantid\scripts\Diffraction\isis_powder\routines\instrument_settings.py", line 124, in _raise_user_param_missing_error
 raise AttributeError(err_text)
```

DannyHindson replied:
OK. I think the offending line needs a `hasattr(self._inst_settings, "pdf_type")` added to the if statement.

The problem I'd tried to fix was that if the pdf type was specified in the confi file or in the line of your script where the Polaris object is created, it was ignored and the default G(r) still kicked in.

RichardWaiteSTFC reviewed (dismissed):
Looks good thanks for the changes. I'm happy to leave the `divide_by_number_of_detectors_in_bank` duplicated for now, it's a small function. Maybe a good new starter issue?

I've tested this again it looks OK - maybe some more structured background features at low-Q in the per-detector reduction - but the FT look very similar (to my untrained eye at least)

DannyHindson committed:
Fix handling if pdf type not supplied

RichardWaiteSTFC reviewed (approved):
Looks good - that fixed the error - now it defaults to `G(r)` if I don't supply an argument in the function call or in the config.

DannyHindson commented:
@mantidproject/gatekeepers - please can someone review\merge this PR? There's another change in this area I'd like to get through by the end of the week
